runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -229,56 +211,64 @@ impl Bank {
+        let new_target_program_data_account = self.new_target_program_data_account(&source)?;
 
         // Gather old and new account data sizes, for updating the bank's
         // accounts data size delta off-chain.
-        // The old data size is the total size of all accounts involved.
-        // The new data size is the total size of the source program accounts,
-        // since the target program account is replaced with a new program
-        // account of the same size as the source program account, and the
-        // source program data account is copied to the target program data
-        // account before both source program accounts are cleared.
-        let target_program_len = target.program_account.data().len();
-        let source_program_len = source.program_account.data().len();
-        let source_program_data_len = source.program_data_account.data().len();
+        // The old data size is the total size of all original accounts
+        // involved.
+        // The new data size is the total size of all the new program accounts.
         let old_data_size = checked_add(
-            target_program_len,
-            checked_add(source_program_len, source_program_data_len)?,
+            target.program_account.data().len(),
+            source.buffer_account.data().len(),
+        )?;
+        let new_data_size = checked_add(
+            new_target_program_account.data().len(),
+            new_target_program_data_account.data().len(),
         )?;
-        let new_data_size = checked_add(source_program_len, source_program_data_len)?;
 
         // Deploy the new target Core BPF program.
         // This step will validate the program ELF against the current runtime
         // environment, as well as update the program cache.
         self.directly_invoke_loader_v3_deploy(
             &target.program_address,
-            &source.program_data_account,
+            new_target_program_data_account.data().len(),
+            &source.buffer_account.data()[BUFFER_METADATA_SIZE..],
+        )?;
+
+        // Calculate the lamports to burn.
+        // The target program account will be replaced, so burn its lamports.
+        // The source buffer account will be cleared, so burn its lamports.
+        // The two new program accounts will need to be funded.
+        let lamports_to_burn = checked_add(
+            target.program_account.lamports(),
+            source.buffer_account.lamports(),
+        )?;
+        let lamports_to_fund = checked_add(
+            new_target_program_account.lamports(),
+            new_target_program_data_account.lamports(),
         )?;
 
-        // Burn lamports from the target program account, since it will be
-        // replaced.
-        self.capitalization
-            .fetch_sub(target.program_account.lamports(), Relaxed);
+        // Update the bank's capitalization.
+        if lamports_to_burn > lamports_to_fund {
```

Comment:
I suppose there's no harm in handling, but can this case ever actually happen? ie. can the target.program_account data ever be longer than 44 bytes (size of program account plus diff of buffer-metadata and program-data metadata)?

Reply:
Good point. You're right this isn't a possible case. A stretch case could be the metadata gets changed in the future and causes that possibility to arise here.